Samples/Server/Directx-SpinningCube/App.cpp

```diff
@@ -350,35 +362,94 @@ bool AppMain(BOOL stopping)
 	}
 
 	// Main loop.
-	MSG msg;
-	BOOL gm;
-	while (!stopping && (gm = ::GetMessage(&msg, NULL, 0, 0)) != 0 && gm != -1)
+	while (!stopping)
 	{
-		// For system service, ignore window and swap chain.
-		if (serverConfig->server_config.system_service)
+		MSG msg = { 0 };
+
+		if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE))
 		{
-			::TranslateMessage(&msg);
-			::DispatchMessage(&msg);
+			if (serverConfig->server_config.system_service ||
+				!wnd.PreTranslateMessage(&msg))
+			{
+				TranslateMessage(&msg);
+				DispatchMessage(&msg);
+			}
 		}
 		else
 		{
-			if (!wnd.PreTranslateMessage(&msg))
+			if (conductor->is_closing())
 			{
-				::TranslateMessage(&msg);
-				::DispatchMessage(&msg);
+				break;
 			}
 
 			if (conductor->connection_active() || client.is_connected())
 			{
-				g_deviceResources->Present();
+				ULONGLONG tick = GetTickCount64();
+				if (!g_deviceResources->IsStereo())
+				{
+					if (g_hasNewInputData)
+					{
+						g_cubeRenderer->Update(g_eyeVector, g_lookAtVector, g_upVector);
+						g_hasNewInputData = false;
+					}
+					else
+					{
+						g_cubeRenderer->Update();
+					}
+
+					// For system service, we render to buffer instead of swap chain.
+					if (serverConfig->server_config.system_service)
+					{
+						g_cubeRenderer->Render(bufferCapturer->GetRenderTargetView());
+						bufferCapturer->SendFrame();
+					}
+					else
+					{
+						g_cubeRenderer->Render();
+						bufferCapturer->SendFrame(frameBuffer.Get());
+						g_deviceResources->Present();
+					}
+
+					// FPS limiter.
+					const int interval = 1000 / nvEncConfig->capture_fps;
+					ULONGLONG timeElapsed = GetTickCount64() - tick;
+					DWORD sleepAmount = 0;
+					if (timeElapsed < interval)
+					{
+						sleepAmount = interval - timeElapsed;
+					}
+
+					Sleep(sleepAmount);
+				}
+				// In stereo rendering mode, we only update frame whenever
+				// receiving any input data.

```

Comment:
wait this sounds like a great optimization, why don't we do this for both stereo and mono?

Reply:
Because in stereo mode, the rendering isn't controlled by fps. It's only updated whenever there is a new input message for camera transform. This is required for frame prediction. On the client side, the input message is sent approximately 60 times per second (using VSync).